lib/logscheduler.c

```diff
@@ -0,0 +1,429 @@
+    {
+      /* our work() function returned right before a new batch was added, let's restart */
+      needs_restart = TRUE;
+    }
+  else
+    partition->flush_running = FALSE;
+
+  g_mutex_unlock(&partition->batches_lock);
+
+  if (needs_restart)
+    main_loop_io_worker_job_submit(&partition->io_job, NULL);
+}
+
+static void
+_partition_add_batch(LogSchedulerPartition *partition, LogSchedulerBatch *batch)
+{
+  gboolean trigger_flush = FALSE;
+
+  g_mutex_lock(&partition->batches_lock);
+  if (!partition->flush_running &&
+      iv_list_empty(&partition->batches))
+    {
+      trigger_flush = TRUE;
+      partition->flush_running = TRUE;
+    }
+  iv_list_add_tail(&batch->list, &partition->batches);
+  g_mutex_unlock(&partition->batches_lock);
+
+  if (trigger_flush)
+    {
+      main_loop_io_worker_job_submit_continuation(&partition->io_job, NULL);
+    }
+
+}
+
+static void
+_partition_init(LogSchedulerPartition *partition, LogPipe *front_pipe)
+{
+  main_loop_io_worker_job_init(&partition->io_job);
+  partition->io_job.user_data = partition;
+  partition->io_job.work = _work;
+  partition->io_job.completion = _complete;
+  partition->io_job.engage = NULL;
+  partition->io_job.release = NULL;
+
+  partition->front_pipe = front_pipe;
+
+  INIT_IV_LIST_HEAD(&partition->batches);
+  g_mutex_init(&partition->batches_lock);
+}
+
+void
+_partition_clear(LogSchedulerPartition *partition)
+{
+  g_mutex_clear(&partition->batches_lock);
+}
+
+/* LogSchedulerThreadState */
+
+static glong
+_get_template_hash(LogTemplate *template, LogMessage *msg)
+{
+  if (log_template_is_literal_string(template))
+    {
+      return g_str_hash(log_template_get_literal_value(template, NULL));
+    }
+  else if (log_template_is_trivial(template))
+    {
+      NVHandle handle = log_template_get_trivial_value_handle(template);
+
+      g_assert(handle != LM_V_NONE);
+      return g_str_hash(log_msg_get_value(msg, handle, NULL));
+    }
+  else
+    {
+      GString *buffer = scratch_buffers_alloc();
+      LogTemplateEvalOptions options = DEFAULT_TEMPLATE_EVAL_OPTIONS;
+
+      log_template_format(template, msg, &options, buffer);
+      return g_str_hash(buffer->str);
+    }
+}
+
+
+static gint
+_get_partition_index(LogScheduler *self, LogSchedulerThreadState *thread_state, LogMessage *msg)
+{
+  if (!self->options->partition_key)
+    {
+      gint partition_index = thread_state->last_partition;
+      thread_state->last_partition = (thread_state->last_partition + 1) % self->options->num_partitions;
+      return partition_index;
+    }
+  else
+    {
+      return _get_template_hash(self->options->partition_key, msg) % self->options->num_partitions;
+    }
+}
+
+static gpointer
+_flush_batch(gpointer s)
+{
+  LogScheduler *self = (LogScheduler *) s;
+
+  gint thread_index = main_loop_worker_get_thread_index();
+  g_assert(thread_index >= 0);
+
+  LogSchedulerThreadState *thread_state = &self->thread_states[thread_index];
```

Comment:
`thread_index` sometimes can be higher than `sizeofarray(thread_states)`.

Reply:
how? the _flush_batch() callback is associated with a specific thread in _queue_thread(), which in turn is called by log_scheduler_push() which validates thread_index against self->num_threads (and against 0).

Reply:
Yeah, I was wrong. Thank you.

I just saw that we use `main_loop_worker_get_thread_index()` without validating against `self->num_threads`.